Transient: merge several notebooks into one

kanglcn opened the pull request:
This update includes:

- Merge base notebook and approach notebooks into only one notebook;
- Follow the secular and co-seismic notebooks to provide AWS downloading;
- Modify the structure of the transient notebook to keep consistency with the other two.

kanglcn commented:
I don't know why the history commits appear here. In fact, I only have two new commits.

yunjunz commented:
> I don't know why the history commits appear here. In fact, I only have two new commits.

`git rebase upstream/main` on your fork will sync your history with the upstream here, then the historical commits will disappear. Here is a detailed tutorial if you are interested: https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history

kanglcn commented:
>

kanglcn commented:
> > I don't know why the history commits appear here. In fact, I only have two new commits.
> 
> `git rebase upstream/main` on your fork will sync your history with the upstream here, then the historical commits will disappear. Here is a detailed tutorial if you are interested: https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history

Thanks @yunjunz . I am pretty interested. I don't know what's the difference between `git rebase` and `git merge` now. Probably I will get it from this tutorial.

EJFielding commented on methods/transient/NISAR_SES_Transient_Requirement_Validation.ipynb:
```diff
@@ -6,99 +6,50 @@
     "tags": []
```
<p>Please add another cell with some text like this to highlight this final result for Approach 1:</p><pre class="ql-syntax" spellcheck="false">&lt;div class="alert alert-warning"&gt;
Final result Method 2—
&nbsp; &nbsp; 68% of points below the requirements line is success
&lt;/div&gt;


</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nisar-solid/ATBD/pull/29/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='94'/>

yunjunz replied:
@kanglcn could you address this suggestion too?

EJFielding commented:
The new combined notebook looks great. I recommend adding a cells to highlight the final results of Approach 1 and Approach 2 in the notebook, as mentioned above.

yunjunz commented:
Correct me if I am wrong please, our agreed naming convention for the notebook is `Transient_Requirement_Validation.ipynb` without `NISAR_SES`, could you rename it to that?

EJFielding commented:
Yes, please remove the "NISAR_SES" part of the name.

kanglcn commented:
<p>Sorry I am busy with some other things. I will do it tomorrow. Thanks.</p>

kanglcn commented:
I have added the result highlight and renamed the transient notebook. But I got conflicts when I did the `git rebase`. Here is the `git status`:
```
rebase in progress; onto fd7f1bd
You are currently rebasing branch 'main' on 'fd7f1bd'.
  (fix conflicts and then run "git rebase --continue")
  (use "git rebase --skip" to skip this patch)
  (use "git rebase --abort" to check out the original branch)

Unmerged paths:
  (use "git restore --staged <file>..." to unstage)
  (use "git add <file>..." to mark resolution)
        both added:      methods/transient/Transient_approach1.ipynb
        both added:      methods/transient/Transient_approach2.1.ipynb
        both added:      methods/transient/Transient_approach2.2.ipynb
        both added:      methods/transient/Transient_base.ipynb

no changes added to commit (use "git add" and/or "git commit -a")
```
I don't know how to fix it now but I will investigate it further. @yunjunz Does it affect our upstream repository here? If so, please help me to fix it. Thanks!

yunjunz commented:
> I don't know how to fix it now but I will investigate it further. @yunjunz Does it affect our upstream repository here? If so, Could you please help me to fix it? Thanks!

The PR won't affect the repo until it's merged, so no worries. I am not sure what was the conflict without more information, but from here things look good.

EJFielding commented:
<p>The site parameters were not exactly the same for all three notebooks (time intervals can be different), but they are almost the same. I suggest keeping this in the notebooks so people know what sites are available.</p>

---

 <a href='https://app.reviewnb.com/nisar-solid/ATBD/pull/29/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

yunjunz commented:
<p>I see. Thanks for the clarification @EJFielding.</p>

---

 <a href='https://app.reviewnb.com/nisar-solid/ATBD/pull/29/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

kanglcn commented:
> > I don't know how to fix it now but I will investigate it further. @yunjunz Does it affect our upstream repository here? If so, Could you please help me to fix it? Thanks!
> 
> The PR won't affect the repo until it's merged, so no worries. I am not sure what was the conflict without more information, but from here things look good.

Great. There is one option to skip these conflicts but we don't need to run the risk since things look good on your side. I will try to skip them after this pull request is merged to check if it works, if not, I will make a new fork from this `nisar-solid/ATBD` repository.

kanglcn commented:
<p>Good to know it. I will do the simplification.</p>

---

 <a href='https://app.reviewnb.com/nisar-solid/ATBD/pull/29/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

kanglcn commented:
<p>As I know, <code>np.savetxt()</code> can only save <code>numpy ndarray</code> but we have the <code>atr</code> which is an <code>dict</code> . I use this <code>pickle</code> just because the downloading of GNUW products takes too much time. Since all of them are staged on s3, I think we don't need to save anything to disk now. Do you agree? If so, I will remove them.     </p>

---

 <a href='https://app.reviewnb.com/nisar-solid/ATBD/pull/29/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

kanglcn commented:
<p>As I know, Jupyterlab will automatically generate a table of contents if proper Markdown title syntax is used. I will send a screenshot to you since I didn't find a way to upload pictures on github and app.reviewnb.</p>

---

 <a href='https://app.reviewnb.com/nisar-solid/ATBD/pull/29/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

EJFielding commented:
<p>Yes, I would go ahead and remove the <code>pickle.dump</code> calls. I already removed them from the Secular and Coseismic notebooks. Starting with the staged data that is already pre-processed into MintPy format means it is fast to regenerate the displacement datasets. </p>

---

 <a href='https://app.reviewnb.com/nisar-solid/ATBD/pull/29/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

kanglcn commented:
<p>Sure.</p>

---

 <a href='https://app.reviewnb.com/nisar-solid/ATBD/pull/29/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

kanglcn commented:
I have updated the transient notebook to follow your suggestions. Please tell me if you have any questions or comments. And @yunjunz , does mintpy have the function to correct the troposphere errors for individual interferograms rather time series results? Thanks!

yunjunz commented:
Yes, `tropo_pyaps3.py` works on single isce2 unwrapped interferogram file also, you could check `tropo_pyaps3.py -h` for examples and details.

yunjunz commented:
<p>Could you merge the above 5 code cells into one?</p>

---

 <a href='https://app.reviewnb.com/nisar-solid/ATBD/pull/29/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

yunjunz commented on methods/transient/NISAR_SES_Transient_Requirement_Validation.ipynb:
```diff
@@ -6,99 +6,50 @@
     "tags": []
```
and here @kanglcn

yunjunz commented on methods/transient/Transient_Requirement_Validation.ipynb:
```diff
@@ -0,0 +1,5142 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #7.    import pickle
</pre><p>This import can be removed too.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nisar-solid/ATBD/pull/29/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

yunjunz replied:
<p>merge the above 2 code cells into one please.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nisar-solid/ATBD/pull/29/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

yunjunz replied:
<p>Use <code>Tides</code> instead of <code>Tide</code> throughout the entire notebook please.  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nisar-solid/ATBD/pull/29/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>

yunjunz commented:
Thank you @kanglcn for the updates, they look great! I put in a few more very minor comments, hopefully the last round! Could you please check the "highlight" suggestions from Eric too?

kanglcn commented:
> Thank you @kanglcn for the updates, they look great! I put in a few more very minor comments, hopefully the last round! Could you please check the "highlight" suggestions from Eric too?

Hi, @yunjunz  I update the transient notebook to fix gramma issues, remove import pickle, merge some cells. I have already highlighted the result as @EJFielding said in my second push. Can you see it at your side?
And, when I use `tropo_pyaps3.py -f ifgramStack.h5 -g geometryGeo.h5` to correct the interferograms (both the ifgramStack.h5 and geometryGeo.h5 is from the `GNUW product download -> ARIA setup -> Mintpy load` pipeline), I get
```
weather model: ERA5 - dry (hydrostatic) and wet delay
weather directory: /home/jovyan/ATBD_kangl/methods/transient/work/transient_ouputs/CentralValleyA144/MintPy/inputs
output tropospheric delay file: ERA5.h5
output corrected time-series file: ifgramStack_ERA5.h5
read dates/time info from file: ifgramStack.h5
Traceback (most recent call last):
  File "/opt/conda/envs/NISAR_SE/bin/tropo_pyaps3.py", line 10, in <module>
    sys.exit(main())
  File "/opt/conda/envs/NISAR_SE/lib/python3.8/site-packages/mintpy/tropo_pyaps3.py", line 834, in main
    read_inps2date_time(inps)
  File "/opt/conda/envs/NISAR_SE/lib/python3.8/site-packages/mintpy/tropo_pyaps3.py", line 217, in read_inps2date_time
    inps.date_list = ptime.yyyymmdd(atr['DATE12'].split('-'))
KeyError: 'DATE12'
```
This error is thrown when loading date from https://github.com/insarlab/MintPy/blob/6e592f8e46ef7e8513e7c1ecce98e85fc278650f/mintpy/tropo_pyaps3.py#L217
It looks like pyaps treat the input interferogram stack as time series otherwise the date should be loaded by https://github.com/insarlab/MintPy/blob/6e592f8e46ef7e8513e7c1ecce98e85fc278650f/mintpy/tropo_pyaps3.py#L807
Do you have any suggestions to use pyaps correctly? Thanks!

yunjunz commented:
@kanglcn For `tropo_pyaps3`, I don't think it supports the ifgramStack file as input, only a single interferogram file (e.g. *.unw) is supported.

You will need the metadata (generated by prep_*.py by default) and referencing it in space before feeding it tropo_pyaps3.py. Here (https://github.com/yunjunz/2021-Maduo-EQ/blob/main/3_prep_ifgram.ipynb) is an example. 

This is not supported in the routine workflow, so you need to get familiar with the individual mintpy stand-alone scripts to get there.

yunjunz reviewed (approved):
Looks all good to me! Thank you @kanglcn!
